packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -3233,10 +3233,12 @@ export default (superClass: Class<Parser>): Class<Parser> =>
 
       // Either way, we're looking at a '<': tt.jsxTagStart or relational.
 
-      let typeParameters: ?N.TsTypeParameterDeclaration;
-      let invalidSingleType: ?N.TsTypeParameter;
-      state = state || this.state.clone();
+      // If the state was cloned in the JSX parsing branch above but there
+      // have been any error in the tryParse call, this.state is set to state
+      // so we still need to clone it.
+      if (!state || state === this.state) state = this.state.clone();
```

Comment:
This change is because the below `tryParse` call should always receive a clone of `this.state`, and not `this.state` itself.